src/collector/Round.cpp

```diff
@@ -145,10 +148,93 @@ void Round::start()
             (m_type == REGULAR) ? "regular" : (m_type == FORCED_FULL) ? "forced full" : "forced partial",
             (m_type == FORCED_PARTIAL ? m_entries.nodes.size() : m_storage->get_nodes().size()));
 
-    dispatch_async_f(m_queue, this, &Round::step2_curl_download);
+    dispatch_async_f(m_queue, this, &Round::step2_1_jobs);
+    dispatch_async_f(m_queue, this, &Round::step2_2_curl_download);
+}
+
+void Round::step2_1_jobs(void *arg)
+{
+    Round & self = *static_cast<Round*>(arg);
+
+    Stopwatch watch(self.m_clock.jobs_database);
+
+    try {
+        const Config & config = self.get_app().get_config();
+
+        if (config.metadata.url.empty() || config.metadata.jobs.db.empty()) {
+            BH_LOG(self.get_app().get_logger(), DNET_LOG_WARNING,
+                    "Not connecting to jobs database because it was not configured");
+            return;
+        }
+
+        std::string errmsg;
+        mongo::ConnectionString cs = mongo::ConnectionString::parse(config.metadata.url, errmsg);
+        if (!cs.isValid()) {
+            BH_LOG(self.get_app().get_logger(), DNET_LOG_ERROR,
+                    "Mongo client ConnectionString error: %s", errmsg.c_str());
+            return;
+        }
+
+        std::unique_ptr<mongo::DBClientReplicaSet> conn((mongo::DBClientReplicaSet *) cs.connect(
+                errmsg, double(config.metadata.options.connectTimeoutMS) / 1000.0));
+        if (conn == nullptr) {
+            BH_LOG(self.get_app().get_logger(), DNET_LOG_ERROR,
+                    "Connection failed: %s", errmsg.c_str());
+            return;
+        }
+
+        std::ostringstream db_ostr;
+        db_ostr << config.metadata.jobs.db << ".jobs";
+
+        mongo::BSONObjBuilder builder;
+        builder.append("id", 1);
+        builder.append("status", 1);
+        builder.append("group", 1);
+        builder.append("type", 1);
+        mongo::BSONObj fields = builder.obj();
+
+        std::auto_ptr<mongo::DBClientCursor> cursor = conn->query(db_ostr.str(),
+                MONGO_QUERY("status" << mongo::NE << "completed"
+                         << "status" << mongo::NE << "cancelled").readPref(
+                             mongo::ReadPreference_PrimaryOnly, mongo::BSONArray()),
+                0, 0, &fields);
+
+        uint64_t ts = 0;
+        clock_get(ts);
+        std::vector<Job> jobs;
+
+        while (cursor->more()) {
+            mongo::BSONObj obj = cursor->next();
+
+            std::string error_text;
+            Job job(ts);
+            if (job.init(obj, error_text) != 0) {
+                BH_LOG(self.get_app().get_logger(), DNET_LOG_ERROR,
+                        "Failed to initialize Job: %s\nBSON object: %s",
+                        error_text.c_str(), obj.jsonString(mongo::Strict, 1));
+                continue;
+            }
+
+            jobs.emplace_back(std::move(job));
```

Comment:
Having used Job's constructor for initialization this code could be rewritten in a bit compact and a slightly more native way:
```
mongo::BSONObj obj = cursor->next();
try {
  jobs.emplace_back(Job{ts, obj});
} catch (const std::exception &ex) {
  BH_LOG(...);
  continue;
}
```